### PR TITLE
feat: call internal functions from constructor

### PR DESCRIPTION
**Pull request by charles-cooper**
### What I did
allow calling internal functions from constructor. fixes https://github.com/vyperlang/vyper/issues/1631 and https://github.com/vyperlang/vyper/issues/2318

### Commit message
```
this commit allows the user to call internal functions from the     
`__init__` function. it does this by generating a call graph during the
annotation phase and then generating code for the functions called from
the init function for during deploy code generation                 
          
this also has a performance benefit (compiler time) because we can get
rid of the two-pass method for tracing frame size.                  
          
now that we have a call graph, this commit also introduces a topsort of
functions based on the call dependency tree. this ensures we can compile
functions that call functions that occur after them in the source code.
          
lastly, this commit also refactors vyper/codegen/module.py so that the
payable logic is cleaner, it uses properties instead of calculations
more, and cleans up properties on IRnode, FunctionSignature and Context.
```

### Cute Animal Picture

image

### CODE REVIEW

**Review by fubuloubu — APPROVED** [review]

**Review by fubuloubu — COMMENTED** [review]
Needs tests and documentation internally talking about the flow that is taken

**Comment by charles-cooper** [comment]
ah i ended up doing some refactoring and forgot to turn this back into a draft

**Comment by charles-cooper** [comment]
btw it now depends on a change to `build_gas_estimates` introduced in https://github.com/vyperlang/vyper/pull/2647

File: examples/stock/company.vy [hunk 1]
```diff
@@ -154,3 +136,21 @@ def debt() -> uint256:
...
+@view
+@internal
+def _stockAvailable() -> uint256:
+    return self.holdings[self.company]
+
+# Find out how much stock any address (that's owned by someone) has.
+@view
+@internal
+def _getHolding(_stockholder: address) -> uint256:
+    return self.holdings[_stockholder]
```
Comment (fubuloubu) [inline examples/stock/company.vy:156]: These two methods aren't really that useful

File: vyper/codegen/function_definitions/common.py [hunk 2]
```diff
@@ -14,91 +13,69 @@
...
         # frame_size of external function includes all private functions called
-        o.total_gas = o.gas + calc_mem_gas(frame_size)
+        o.add_gas_estimate += calc_mem_gas(sig.frame_info.mem_used)
     else:
-        # frame size for internal function does not need to be adjusted
-        # since it is already accounted for by the caller
-        o.total_gas = o.gas
+        # note: internal functions do not need to adjust gas estimate since
+        # it is already accounted for by the caller.
+        pass
```
Comment (fubuloubu) [inline vyper/codegen/function_definitions/common.py:77]: remove `else` case

File: examples/stock/company.vy (hunk 1 above)
Comment (charles-cooper) [reply examples/stock/company.vy:156]: true, i think they are good there just to demonstrate that internal functions can be defined after they are called

**Review by fubuloubu — APPROVED** [review]

**Pull request merged**